File: matlab/arnoldi/arnoldi__orth_stabilitymetric.m
%
function [ repres, orth, condn, normS ] = arnoldi__orth_stabilitymetric( A, b, Q, H, beta )
%
   [m] = size(A,1);
   [k] = size(H,1);
%
   if( (size(A,1) ~= m) & (size(A,2) ~= m) )   fprintf("ERROR1 in arnoli__orth_stabilitymetric\n"); end;
   if( (size(b,1) ~= m) & (size(b,2) ~= 1) )   fprintf("ERROR2 in arnoli__orth_stabilitymetric\n"); end;
   if( (size(Q,1) ~= m) & (size(Q,2) ~= k) )   fprintf("ERROR3 in arnoli__orth_stabilitymetric\n"); end;
   if( (size(H,1) ~= k) & (size(H,2) ~= k-1) ) fprintf("ERROR4 in arnoli__orth_stabilitymetric\n"); end;
%
   repres = zeros(1,k);
   orth   = zeros(1,k);
   condn  = zeros(1,k);
   normS  = zeros(1,k);
%
   nrmA = norm(A,'fro');
%
   for j = 1:k,
%
      orth(j)   = norm( eye(j) - Q(1:m,1:j)' * Q(1:m,1:j), 'fro');
%
      repres(j) = norm( [ b, A * Q(1:m,1:j-1) ] - Q(1:m,1:j) * [ [beta; zeros(j-1,1) ], H(1:j,1:j-1)], 'fro') / nrmA;
%
      condn(j)  = cond( [ b, A * Q(1:m,1:j-1) ] );
%
      U(1:j,1:j) = triu(Q(1:m,1:j)' * Q(1:m,1:j), 1);
      S = (eye(j) + U) \ U;
      normS(j) = norm(S,2);
%
   end
%
end
%

File: julia/arnoldi/arnoldi__orth_stabilitymetric.jl
using LinearAlgebra

function arnoldi__orth_stabilitymetric(A, b::AbstractVector,
                                        Q::AbstractMatrix, H::AbstractMatrix,
                                        beta)
    m = size(A, 1)
    k = size(H, 1)
    # orth methods return beta as a scalar or a 1-element vector
    β = isa(beta, AbstractVector) ? beta[1] : beta

    nrmA   = norm(reshape(A, :))
    repres = zeros(k)
    orth   = zeros(k)
    condn  = zeros(k)
    normS  = zeros(k)

    for j in 1:k
        Qj = Q[:, 1:j]
        orth[j]   = norm(reshape(Matrix(I, j, j) .- Qj' * Qj, :))
        repres[j] = norm(hcat(b, A * Q[:, 1:j-1]) .- Qj * hcat([β; zeros(j-1)], H[1:j, 1:j-1])) / nrmA
        condn[j]  = cond(hcat(b, A * Q[:, 1:j-1]))
        U = triu(Qj' * Qj, 1)
        S = (Matrix(I, j, j) .+ U) \ U
        normS[j] = opnorm(S, 2)
    end

    return repres, orth, condn, normS
end
